### PR TITLE
Stm version 1

### PR DESCRIPTION
**Pull request by RubelMozumder**
This PR implies:

1. **STM and STS reader**: At this the reader supports two version (Generic 5e, Generic 4.5) of STM and STS experiment (from Sabine and Carlos).
2. **generate_eln** tools with option **--eln-type (allowed values eln, echem_eln)** for simple eln and scheme eln.
3. nexus converter checks the existence of the concept that is referred to be linked.

### CODE REVIEW

**Comment by RubelMozumder** [comment]
The STS reader is already updated in Nomad. Once this PR will be accepted, the corresponding branch in nomad will be update and ready to merge into develop branch.

File: pynxtools/eln_mapper/schema_eln.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+#
+
+from typing import Dict, Any
+import xml.etree.ElementTree as ET
+import yaml
+from pynxtools.eln_mapper.eln import retrieve_nxdl_file
+from pynxtools.dataconverter.helpers import remove_namespace_from_tag
+
+
+NEXUS_TYPE_TO_NOMAD_TYPE = {'NX_CHAR': {'convert_typ': 'str',
```
Comment (mkuehbach) [inline pynxtools/eln_mapper/schema_eln.py:27]: better rename to NEXUS_TYPE_TO_NUMPY_TYPE rather than NOMAD_TYPE because an np.float64 etc. is nothing specific for NOMAD

File: pynxtools/eln_mapper/schema_eln.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                                     'sections': {}}
+    sections = recursive_dict['definitions']['sections']
+
+    scan_xml_element_recursively(nxdl_root, sections,
+                                 root_name=root_name, is_root=True)
+
+
+def generate_schema_eln(nexus_def: str, eln_file: str = None) -> None:
+    """Generate schema eln that should go to Nomad while running
+    the reader. The output file will be <eln_file>.scheme.archive.yaml
```
Comment (mkuehbach) [inline pynxtools/eln_mapper/schema_eln.py:240]: schema.archive.yaml not scheme.archive.yaml

File: pynxtools/eln_mapper/schema_eln.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    root_name = nxdl_root.attrib['name'][2:] if 'name' in nxdl_root.attrib else "<ROOT_NAME>"
+    recursive_dict['definitions'] = {'name': '<ADD PREFERED NAME>',
+                                     'sections': {}}
+    sections = recursive_dict['definitions']['sections']
+
+    scan_xml_element_recursively(nxdl_root, sections,
+                                 root_name=root_name, is_root=True)
+
+
+def generate_schema_eln(nexus_def: str, eln_file: str = None) -> None:
```
Comment (mkuehbach) [inline pynxtools/eln_mapper/schema_eln.py:238]: eln_file_name I've seen that you name functions and variables already often good to indicate what is the intention behind these I think you can do this even more succinct, more a suggestions for the future

File: pynxtools/dataconverter/readers/stm/reader.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+
+from typing import Any, Dict, Tuple, Union
+from collections.abc import Callable
+import json
+import yaml
+
+from pynxtools.dataconverter.readers.base.reader import BaseReader
+from pynxtools.dataconverter.template import Template
+from pynxtools.dataconverter.readers.stm.sts_file_parser import from_dat_file_into_template
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/reader.py:30]: I see STS seems some dat file format while STM is nanonis, is the dat file generated from a different instrument?
I am asking this as I think it is useful to document in the source code that even though these are just two examples they follow the concept for users,
hej you are a scientist working in STM/STS, we know there are many different formats and different instruments: here is one common concept exemplified for these specific two types of combinations (machine A + file format A) and (machine B + file format B) that sort of story

File: pynxtools/dataconverter/readers/stm/reader.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    'Environment': 'ENVIRONMENT[environment]',
+    'Sample_bias': 'SAMPLE_BIAS[sample_bias]'
+}
+
+REPLACE_NESTED: Dict[str, str] = {}
+
+
+# pylint: disable=too-few-public-methods
+class StmNanonisGeneric5e:
+    """Class to handle 'stm' experiment of software version 'Generic 5e' from 'nanonis'
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/reader.py:56]: okay now I understand that for now all your stm is specifically for nanonis software, generic 5e

File: pynxtools/dataconverter/readers/stm/stm_helper.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+# limitations under the License.
+
+from typing import Tuple
+import json
+import numpy as np
+from pynxtools.dataconverter.helpers import convert_data_dict_path_to_hdf5_path
+
+
+# Here are some data or data type or unit or data to skip:
+UNIT_TO_SKIP = ['on/off', 'off', 'on', 'off/on']
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_helper.py:27]: map to boolean instead, I have seen such example also from mpes, could be moved into a helper for the converter to be shared across all pynxtools readers

File: pynxtools/dataconverter/readers/stm/stm_helper.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+def slice_before_last_element(np_array):
+    """Get all the elements before last element.
+    """
+    if not isinstance(np_array, np.ndarray) and not len(np.shape(np_array)) == 1:
+        raise ValueError('Please provide a numpy array of 1D.')
+    return np_array[:-1]
+
+
+# pylint: disable=too-many-return-statements
+def transform(str_value):
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_helper.py:171]: What is being transformed? The name of this function is too general.

File: pynxtools/dataconverter/readers/stm/stm_helper.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+            if orig_link_path in link_modified_dict:
+                # modified concepts come in a list together.
+                modif_link_hdf_path = convert_data_dict_path_to_hdf5_path(
+                    link_modified_dict[orig_link_path])
+                val['link'] = modif_link_hdf_path
+            else:
+                val['link'] = convert_data_dict_path_to_hdf5_path(orig_link_path)
+
+
+def cal_dx_by_dy(x_val: np.ndarray, y_val: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_helper.py:146]: I would in general move such scientific processing even though it is super simple (for now) to an own source code file, to keep the helper code as structured right from the start as possible

File: pynxtools/dataconverter/readers/stm/stm_helper.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    """Calc conductance or gradiant dx/dy for x-variable and y-variable also return the result."""
+    dx_ = x_val[0::2] - x_val[1::2]
+    dy_ = y_val[0::2] - y_val[1::2]
+
+    dx_by_dy = dx_ / dy_
+
+    return dx_by_dy
+
+
+def cal_x_multi_x(x_val: np.ndarray, y_val: np.ndarray) -> np.ndarray:
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_helper.py:156]: Maybe here an in the next function a function call is overkill or are you expecting that in the future the code in each function will get longer/more?

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
+"""
+    A parse stm file in nexus generated definition template.
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_file_parser.py:2]: Please try to force you to make as clean as possible docstrings, I know also that I frequently make typo and grammar mistakes but we should try our best it really adds to code readability and interpretability, I once had a colleague who wrote comments in the code complaining about the functionality of certain code portions in the mother language that inside some Fortran code :) this told me to better when I am old, in case you ask: no I was not that person complaining the source code in my mother tongue :)

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import os
+from typing import Any, Dict
+import logging
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_file_parser.py:25]: This logger I remember @sherjeelshabih rightly so made once the comment in one the readers I wrote that we should use one of the official loggers so that nomad can pick up issues and problems in how far is this realized here? Maybe a topic for the future?

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+                    else:
+                        spreaded_dict[l_key]: Dict[str, Any] = {}
+                        spreaded_dict[l_key] = self.get_nested_dict_from_concatenated_key(
+                            {r_key: d_val}, dict_to_map_path=spreaded_dict[l_key])
+            else:
+                spreaded_dict[d_key] = d_val
+
+        return spreaded_dict
+
+    def convert_key_to_unit_and_entity(self, key, val, start_bracket='', end_bracket=''):
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_file_parser.py:98]: I think one could still cleaner separate the different (currently two (sub-)parsers in this example, for a first round I think what we have lgtm though @RubelMozumder

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+                for tup in tuple_li:
+                    flattened_dict[tup[0]] = tup[1]
+            else:
+                flattened_dict[key] = val
+
+        flattened_dict.update(temp_flattened_dict_sig)
+        return flattened_dict
+
+    # pylint: disable=too-many-arguments
+    def construct_nxdata_for_sxm(self,
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_file_parser.py:175]: I have a general question about this design here as I can pick it up too quickly from the code here:
In general for each file format e.g. nanonis or sxm you have on the left hand side some information
which you want to map on a NeXus concept. What I don't see immediately transparent from the code
where is the logic which decides which of the fields on the left hand side will for sure be copied over and which optionally.
In other words is the code logic reasonably stable to deal with changes of the NXsts appdef when in this appdef required and optional constraints are changed. It seems to me that one strength of your implementation is the fact that you do not directly hardcode paths but try to resolve many paths programmatically which is good.

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+        """
+        Extract dimension info from scanfield.
+
+            ../ENVIRONMENT[environment]/scan_control/positioner/scanfield"
+            The scanfield has four parts starting point of (x, y) co-ordinate
+            length on (x, y)-dimenstion and one last unknown values.
+        """
+        scanfield: str = ''
+        for key, val in config_dict.items():
+            if ("/ENTRY[entry]/INSTRUMENT[instrument]/ENVIRONMENT[environment]/"
```
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_file_parser.py:293]: But this line for instance shows that the left to right mapping (domain file format to NeXus mapping) is maybe quite stiff still because what if we have an ENTRY[entry2] can this be dealt with?
I think for a first round of the parser this is all good...

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py
Comment (mkuehbach) [inline pynxtools/dataconverter/readers/stm/stm_file_parser.py:1]: I think it is indeed useful (like for atom probe and many other examples) to offload the actual file format technical reading into an own module/library which removes the specific file format formatting quirks so that the pynxtools reader can easier I with less format-handling python code read these formats this would clean the code. For the first round though lgtm

**Review by mkuehbach — CHANGES_REQUESTED** [review]
Dear @RubelMozumder,

Congratulations on your first successful parser and example! Overall I think this is a solid amount of work. There are some comments which have to be dealt with, many of the other comments are more suggestions for the future. Maybe you also make a list for these like what Florian and Lukas did for MPES so that the next SPM round of readers can become even better.

One general observation I made is that you should at least for function names and docstrings try to check for the language, I know this is tricky and we all have to fight with it as we're all not English native speakers but I think it is also a great change to professionalize ones language usage which is always useful

Other than this I think your work is a good addition to the pynxtools and will likely find the appreciation of the SPM community

**Comment by RubelMozumder** [comment]
> Dear @RubelMozumder,
> 
> Congratulations on your first successful parser and example! Overall I think this is a solid amount of work. There are some comments which have to be dealt with, many of the other comments are more suggestions for the future. Maybe you also make a list for these like what Florian and Lukas did for MPES so that the next SPM round of readers can become even better.
> 
> One general observation I made is that you should at least for function names and docstrings try to check for the language, I know this is tricky and we all have to fight with it as we're all not English native speakers but I think it is also a great change to professionalize ones language usage which is always useful
> 
> Other than this I think your work is a good addition to the pynxtools and will likely find the appreciation of the SPM community

At some point you are right. In the development process at some point, the name and docs for functions and classes lose their significance due to reshuffling, editing, splitting, and reorganizing of the code. However, I will keep your advice in mind.

File: dev-requirements.txt
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply dev-requirements.txt:1]: LGTM! The dev-requiremts.txt has a new dependency namely **nanonispy**, so the dev-requirements.txt needed to be regenerated. Regarding this change, the requiremts.txt and dev-requirements.txt were also generated in noamd, and no conflict either from nomad, or third-party tools.

File: examples/stm/README.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
+# STM Example
+Several examples regarding several versions of Nanonis instrument are available in [https://gitlab.mpcdf.mpg.de](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-remote-tools-hub/-/tree/develop/docker/stm). But to get in that example one must have an account in https://gitlab.mpcdf.mpg.de. Stll you need to try the example stm reader out, please reach out [Rubel Mozumder](mozumder@physik.hu-berlin.de) or docker container (discussed below).
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply examples/stm/README.md:2]: Zeodo LGTM! I will keep it noted in a newly created issue.

File: pynxtools/dataconverter/readers/stm/README.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    !dataconverter \
+    --reader stm \
+    --nxdl NXsts \
+    --input-file ./STS_nanonis_generic_5e_1.dat \
+    --input-file ../config_file_for_dat.json \
+    --input-file Nanonis_Eln.yaml \
+    --output ./final_sts_dev.nxs
+    ```
+
+- Utilization of ELN
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/dataconverter/readers/stm/README.md:49]: I do not have any plan to keep the example files in Pynxtools. Because they are already in nomad-remote-tools-hub and both nomad-remote-tool-hub and pynxtools both are submodules in  nomad.

File: pynxtools/dataconverter/readers/stm/README.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+                      unit: m
+                      m_annotations:
+                        eln:
+                          component: NumberEditQuantity
+                          defaultDisplayUnit: m
+                      description: |
+                        The scanning area in x position in the frame. (e.g. 130.5E-9).
+
+        ```
+## Config file:
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/dataconverter/readers/stm/README.md:114]: The same reason as before. Actually, the config files does not make any sense for anonymous visitors, rather they are meaningful for users. At the moment the users, I have they are using different labs (from China and HU). So, different labs define different parameters in different ways. Therefore, only they need to change the config files according to their wish. And, the README file would add value for them not for general people.

File: pynxtools/dataconverter/readers/stm/reader.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
+"""
+    A short description on SPM reader.
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/dataconverter/readers/stm/reader.py:2]: I aggree to keep consistency for all reader name.

Personally, I would prefer the SPM reader, as the STM, STS, and AFM belong to the SPM experiment and at this moment the reader supports both STM and STS experimental data files. One essential difference between STM and STS is the area of scan, where STM scans almost the entire sample area and STS in a very narrower area.

File: pyproject.toml [hunk 19]
```diff
@@ -66,6 +68,7 @@ dev = [
 read_nexus = "pynxtools.nexus.nexus:main"
 dataconverter = "pynxtools.dataconverter.convert:convert_cli"
 nyaml2nxdl = "pynxtools.nyaml2nxdl.nyaml2nxdl:launch_tool"
+generate_eln = "pynxtools.eln_mapper.eln_mapper:get_eln"
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pyproject.toml:71]: **get_eln** is just a function name that returns the final ELN files.

File: pyproject.toml [hunk 20]
```diff
@@ -41,6 +41,8 @@ dependencies = [
     "lark>=1.1.5",
     "requests",
     "requests_cache",
+    "nanonispy>=1.1.0",
+    "coverage==6.5.0"
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pyproject.toml:45]: If it is removed, the newest version of **coverage** has bugs and generates errors. While generating **dev-requirement.txt** it collects the newest version of **coverage** and returns errors. So, if **coverage** brings another newer version of bug-free, we can remove it later.

File: pynxtools/dataconverter/template.py [hunk 21]
```diff
@@ -175,8 +175,12 @@ def __delitem__(self, key):
         if key in self.optional.keys():
             del self.optional[key]
 
-        if key in self.required.keys():
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/dataconverter/template.py:178]: I would say it would not generate any side-effect,if nobody delete any required concept. In case, the required concept is deleted the validation of the converter validation step. As our converter validates each concept by backtracking according to the application definition.

File: pynxtools/eln_mapper/schema_eln.py (hunk 2 above)
Comment (RubelMozumder) [reply pynxtools/eln_mapper/schema_eln.py:240]: I will check it again.

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py (hunk 11 above)
Comment (RubelMozumder) [reply pynxtools/dataconverter/readers/stm/stm_file_parser.py:25]: I fully support your point of view. Indeed, I started to think about that but I saw that the repo needs to be restored again.

File: pynxtools/dataconverter/readers/stm/stm_file_parser.py (hunk 13 above)
Comment (RubelMozumder) [reply pynxtools/dataconverter/readers/stm/stm_file_parser.py:175]: This parse only read the numerical data and arrays, collects the units, long-name, that's it. And it fills the concept with the corresponding data. To check about the concepts of optional and required fields passed the converter validation steps. The file parser or stm reader does not have anything to do with required and optional concepts.

**Comment by RubelMozumder** [comment]
Hey @mkuehbach, I think this pull request is ready to merge.

File: pynxtools/eln_mapper/schema_eln.py (hunk 2 above)
Comment (mkuehbach) [reply pynxtools/eln_mapper/schema_eln.py]: What did the check brought about?

**Review by mkuehbach — APPROVED** [review]
Suggest to approve it for now in the interest to move on but fixing typos and suggested refactoring is strongly advised in new versions of this parser.

File: pynxtools/dataconverter/writer.py [hunk 22]
```diff
@@ -287,6 +293,10 @@ def add_units_key(dataset, path):
 
                 if entry_name[0] != "@":
                     path_hdf5 = helpers.convert_data_dict_path_to_hdf5_path(path)
+                    # Handling links that does not exist in hdf file
+                    if isinstance(value, dict) and path_hdf5 not in self.output_nexus:
+                        continue
+
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/writer.py:299]: This can be removed and instead implemented in line: 281

```python
if dataset == None:
   del self.data[path]
```

**Comment by sherjeelshabih** [comment]
I just realized that 
```python 
def rename_entry 
``` 
in dataconverter/template.py also needs to rename the target for links defined in the NXDL.

File: pynxtools/dataconverter/writer.py [hunk 23]
```diff
@@ -279,6 +279,9 @@ def add_units_key(dataset, path):
                               f"with the following message: {str(exc)}") from exc
         for links in hdf5_links_for_later:
             dataset = handle_dicts_entries(*links)
+            if dataset is None:
+                # If target of a link is invalid to be linked
+                del self.data[links[-1]]
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/writer.py:284]: Shouldn't this delete the current path we are checking i.e. the path that holds the links dict in the template not the target? What is links[-1] return?

**Review by mkuehbach — APPROVED** [review]
lgtm

File: pynxtools/dataconverter/readers/sts/__init__.py [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+# To mvake the functions available in stm module
+# pylint: disable=unused-import
+from pynxtools.dataconverter.readers.sts.sts_file_parser import get_sts_raw_file_info
+from pynxtools.dataconverter.readers.sts.stm_file_parser import get_stm_raw_file_info
```
Comment (sanbrock) [inline pynxtools/dataconverter/readers/sts/__init__.py:29]: the reader name is still sts. Do we support sts and/or stm. Is sts the best choice for the name of the reader?
